src/Reopt/TypeInference/Constraints.hs

```diff
@@ -284,11 +346,16 @@ upperBound type1 type2 =
     (s, t@(VarTy _)) -> orTy [s,t]
     -- J-Ptr (N.B., TIE uses ⊓ in this rule, but I think that's a typo)
     ((PtrTy w1 s), (PtrTy w2 t)) | w1 == w2 -> ptrTy w1 (upperBound s t)
+    -- J-RecBase
+    (t1, t2@RecTy{}) | isBaseTy t1 -> upperBound (recTy' [(0,t1)]) t2
+    (t1@RecTy{}, t2) | isBaseTy t2 -> upperBound t1 (recTy' [(0,t2)])
```

Comment:
It seems like in practice the RHS will often (always?) hit `TopTy` here?

Reply:
I don't yet follow, can you say more? Are we talking about specifically the record cases? Or the entirety of `upperBound`? Can you give an example of the sorts of input that fall into this pattern you're concerned about?

Reply:
I just meant, if you follow where these two recursive calls will go, won't they end up in the very last branch and return `TopTy`?

Reply:
Ah, I see. Especially since subtyping has a similar case... yes you're right I think. For `lowerBound` where we're sort of calculating the intersection, pairwise recursing into the fields makes a lot of sense.

For `upperBound` (the function here) when there's two records, we _could_ instead create a union (non-lossy, but might create more unions than we want... hard to know) or we could union the fields, which _is_ lossy... I'm not sure how undesirable that would be in practice (or how often it would occur).... hmmm....